crates/exex/exex/src/notifications.rs

```diff
@@ -13,101 +13,179 @@ use std::{
+        let current = std::mem::replace(&mut self.inner, ExExNotificationsInner::Invalid);
+        self.inner = ExExNotificationsInner::WithHead(match current {
+            ExExNotificationsInner::Invalid => unreachable!(),
+            ExExNotificationsInner::WithoutHead(notifications) => {
+                notifications.with_head(exex_head)
+            }
+            ExExNotificationsInner::WithHead(notifications) => notifications,
+        });
+    }
+
+    /// Returns a new [`ExExNotifications`] with the provided head.
     ///
-    /// For full documentation, see [`Receiver::poll_recv`].
-    #[deprecated(
-        note = "use `ExExNotifications::poll_next` and its `Stream` implementation instead"
-    )]
-    pub fn poll_recv(&mut self, cx: &mut Context<'_>) -> Poll<Option<ExExNotification>> {
-        self.notifications.poll_recv(cx)
+    /// See the documentation of [`ExExNotificationsWithHead`] for more details.
+    pub fn with_head(mut self, exex_head: ExExHead) -> Self {
+        self.inner = ExExNotificationsInner::WithHead(match self.inner {
+            ExExNotificationsInner::Invalid => unreachable!(),
+            ExExNotificationsInner::WithoutHead(notifications) => ExExNotificationsWithHead::new(
+                notifications.node_head,
+                notifications.provider,
+                notifications.executor,
+                notifications.notifications,
+                notifications.wal_handle,
+                exex_head,
+            ),
+            ExExNotificationsInner::WithHead(notifications) => ExExNotificationsWithHead::new(
+                notifications.node_head,
+                notifications.provider,
+                notifications.executor,
+                notifications.notifications,
+                notifications.wal_handle,
+                exex_head,
+            ),
+        });
+        self
     }
 }
 
-impl<P, E> ExExNotifications<P, E>
+impl<P, E> Stream for ExExNotifications<P, E>
 where
     P: BlockReader + HeaderProvider + StateProviderFactory + Clone + Unpin + 'static,
     E: BlockExecutorProvider + Clone + Unpin + 'static,
 {
-    /// Subscribe to notifications with the given head. This head is the ExEx's
-    /// latest view of the host chain.
-    ///
-    /// Notifications will be sent starting from the head, not inclusive. For
-    /// example, if `head.number == 10`, then the first notification will be
-    /// with `block.number == 11`. A `head.number` of 10 indicates that the ExEx
-    /// has processed up to block 10, and is ready to process block 11.
-    pub fn with_head(self, head: ExExHead) -> ExExNotificationsWithHead<P, E> {
+    type Item = eyre::Result<ExExNotification>;
+
+    fn poll_next(
+        self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Option<Self::Item>> {
+        match &mut self.get_mut().inner {
+            ExExNotificationsInner::Invalid => unreachable!(),
```

Comment:
branches are checked in the order they appear, since this is unreachable, this should go last

Reply:
good point, fixed